### ISSUE TITLE
Twisted test suite requires too much memory

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2275]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2275|
|Type|defect|
|Created|2006-12-06 15:34:54Z|

This is finally getting to the point where it is a practical issue.  There are buildslaves which cannot reliable get through the entire test suite without running into allocation failures.

This is a tracking ticket for specific issues which result in an unnecessarily large amount of memory to be allocated by the test suite.

Please add comments related to that topic here.

<details><summary>Searchable metadata</summary>

```
trac-id__2275 2275
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1165419294000000 1165419294000000
changetime__1187714566000000 1187714566000000
version__None None
owner__ 
cc__jml cc__radix cc__itamarst cc__exarkun cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_2"></a>|@radix commented|
|-|-|

Given that we can't trust that the tests are actually passing now, this needs to be fixed before release.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Can you explain this further?  The description of this ticket states that the buildslaves encounter allocation "failures" - the word "failure" here implying to me that the suite fails as a result.  Are you saying that we are getting false positives from buildbot on test runs which have actually aborted due to an allocation failure?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

```
(13:09:43) exarkun: so I don't think there's really a release blocking issue here
(13:10:49) exarkun: it's just one new kind of intermittent failure in the test suite
```

... and I agree, so, removing the milestone.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @therve|
|-|-|

I had a look at this issue. The following diff solves **a lot** of problems:

```
Index: twisted/trial/test/test_tests.py
===================================================================
--- twisted/trial/test/test_tests.py    (revision 20638)
+++ twisted/trial/test/test_tests.py    (working copy)
@@ -47,7 +47,7 @@


     def assertSuccessful(self, test, result):
-        self.assertEqual(result.successes,  [(test,)])
+        self.assertEqual(result.successes,  [(str(test),)])
         self.assertEqual(result.failures, [])
         self.assertEqual(result.errors, [])
         self.assertEqual(result.expectedFailures, [])
Index: twisted/trial/runner.py
===================================================================
--- twisted/trial/runner.py     (revision 20638)
+++ twisted/trial/runner.py     (working copy)
@@ -143,10 +143,12 @@
         """
         # we implement this because Python 2.3 unittest defines this code
         # in __call__, whereas 2.4 defines the code in run.
-        for test in self._tests:
+        while self._tests:
+            test = self._tests.pop(0)
             if result.shouldStop:
                 break
             test(result)
+            del test
         return result


Index: twisted/trial/reporter.py
===================================================================
--- twisted/trial/reporter.py   (revision 20638)
+++ twisted/trial/reporter.py   (working copy)
@@ -144,7 +144,7 @@

         @type test: L{pyunit.TestCase}
         """
-        self.successes.append((test,))
+        self.successes.append((str(test),))

     def upDownError(self, method, error, warn, printStatus):
         pass
```

Is that an acceptable solution? I just fixed the existing tests, but I can add new ones, especially for the ```TestSuite``` change.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I thought about making a similar change recently.  I think the only real problem with that approach is that it breaks -u (I think).  Somebody needs to take responsibility for re-initializing self._tests if the suite is run again.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

Yes, you're right, I hadn't spot that (there's probably a test missing somewhere, because the trial suite pass).

The problem if you re-initialize self._tests is that you change the semantic of until failure. Maybe that's not so different, I don't know. We can also use a different suite for until-failure. After all, it's not used on the buildbots.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [20752]) Add 2 optimizations, fix the tests.

Refs [#2275](https://github.com/twisted/twisted/issues/2275)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

With the above commit, ```trial twisted.conch``` takes 35M of memory instead of 137M before.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [20761]) Add a few tests.

Refs [#2275](https://github.com/twisted/twisted/issues/2275)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner|
|-|-|

I think that's clean enough to review in trial-memory-2275.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_12"></a>|@radix set owner to @therve|
|-|-|

[1] The str(test) change is really weird. I looked into it and I don't even see why you should be keeping any of these as test cases *or* strings. Why not just use a counter?

[2] The whitebox test seems insufficient to me; I suggest instead writing a test that grabs a weakref to the TestCase instance during its run and then after trial has a chance to clean it up, call gc.collect() and make sure the weakref is invalid.

Thanks for the work!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

(In [20875]) Change successes to be a counter.

Refs [#2275](https://github.com/twisted/twisted/issues/2275).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [20876]) Add a blackbox test on the collection of test intances.

Refs [#2275](https://github.com/twisted/twisted/issues/2275)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @radix|
|-|-|

Thanks for the inputs, it should be OK.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst set owner to @therve|
|-|-|

The other categories of test results (failures, errors) can also cause memory leaks; arguably a lot more in the case of errors and failures if the traceback doesn't get converted to string immediately, which I think is what happens now.

So why not, instead of storing test instances and then rendering them with a formatter method, store the result of calling the formatter method? In virtually all cases that should use less memory.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

Replying to [itamarst](#note_16):

> So why not, instead of storing test instances and then rendering them with a formatter method, store the result of calling the formatter method? In virtually all cases that should use less memory.

The problem with that approach is that you limit what is made available to the reporter at the end: if you want further informations about the test that fail, you need the full test object, not just a string representation. Of course for now we don't use that, so that may not matter. My other point is that you don't save many things with that, since there are much more successes than failures (in practice).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_18"></a>|@itamarst commented|
|-|-|

The formatters on the reporter would be called as the tests run, so presumably different reporters would have different formatters?

But it's true there's (usually) less failures than successes, at least in the buildbot scenario.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve commented|
|-|-|

Oh you're right, a Reporter is also a TestResult, so that works. I'm working on it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve commented|
|-|-|

Hum some tests like ```test_tests.TestSkipMethods.test_reasons``` suppose that it got the full test object in skips, so that actually doesn't work. But I fixed some errors due to the change to successes in the process.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_22"></a>|@jml set owner to @therve|
|-|-|

Thanks for doing this.

This branch really makes two independent changes: storing 'successes' as a counter rather than a list, and changing to use `CleanupTestSuite`. It would be easier to review if the changes were in separate branches.

 * Please rename `CleanupTestSuite` to `DestructiveTestSuite`. Trial already uses the word "cleanup" too much, and it should be painfully obvious that this suite destroys data.
 * I really dislike using `assertTrue` and `assertFalse` in `CleanupTestSuiteTestCase.test_cleanup`. Use the `countTestCases` public method instead.
 * That you added `TestSuiteUsed.test_untilFailureSuite` worries me. Did no other test fail before you added the condition in `twisted/scripts/trial.py`?

Other than that, the branch looks good. Please make changes / answer questions and assign back to me.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_23"></a>|@jml commented|
|-|-|

Also, it would be very good to have some measurements of memory usage so we can figure out how effective this branch is.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve commented|
|-|-|

(In [21055])  
 * Rename CleanupTestSuite to DestructiveTestSuite
 * Clean a test

Refs [#2275](https://github.com/twisted/twisted/issues/2275)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_25"></a>|@therve set owner to @jml|
|-|-|

Replying to [jml](#note_22):
> Thanks for doing this.
>
> This branch really makes two independent changes: storing 'successes' as a counter rather than a list, and changing to use `CleanupTestSuite`. It would be easier to review if the changes were in separate branches.

Yep sorry. The changes were intended to be small, but it always grows up.

>  * Please rename `CleanupTestSuite` to `DestructiveTestSuite`. Trial already uses the word "cleanup" too much, and it should be painfully obvious that this suite destroys data.

Done

>  * I really dislike using `assertTrue` and `assertFalse` in `CleanupTestSuiteTestCase.test_cleanup`. Use the `countTestCases` public method instead.

Done.

>  * That you added `TestSuiteUsed.test_untilFailureSuite` worries me. Did no other test fail before you added the condition in `twisted/scripts/trial.py`?

Nope :/. I didn't find any other clean way to do that.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_26"></a>|@therve commented|
|-|-|

Replying to [jml](#note_23):
> Also, it would be very good to have some measurements of memory usage so we can figure out how effective this branch is.

OK, here are some stats, for some suites:

twisted.web:
 * before: VmPeak: 39668 kB, VmSize: 38584 kB
 * after: VmPeak: 30652 kB, VmSize: 28756 kB

twisted.conch:
 * before: VmPeak: 139204 kB, VmSize:139204 kB
 * after: VmPeak: 18944 kB, VmSize: 18944 kB

twisted.test:
 * before: VmPeak: 163504 kB, VmSize: 118104 kB
 * after: VmPeak: 147488 kB, VmSize: 100540 kB

I didn't make a run on whole suite, but I guess it's there where the improvement is the best.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set owner to @therve|
|-|-|

Just a couple things:

  * TestResult.successes is undocumented
  * I wonder if we care about how this changes the behavior of countTestCases - returning an ever diminishing number over the course of a run seems like it might be bad.  trial itself only ever calls it once, at the beginning of a run, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

Hrm.  I guess one other thing.  I'm not sure the single unit test for DestructiveTestSuite is really sufficient.  The only thing it verifies is that there are no test cases left at the end of the run.  There's no verification that the implementation actually runs any of the tests, for example.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_29"></a>|@therve commented|
|-|-|

(In [21113]) 
 * Add doc for successes
 * Add a small test

Refs [#2275](https://github.com/twisted/twisted/issues/2275)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_30"></a>|@therve removed owner|
|-|-|

I think the behavior of countTestCases is sane, and adjusted other things.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun set owner to @therve|
|-|-|

The shouldStop check in the loop is still untested.  But... as far as I can tell, nothing in trial ever sets shouldStop to anything.  Maybe just delete that code?  The rest looks great.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_34"></a>|@therve commented|
|-|-|

(In [21137]) 
 * Add a test for shouldStop

Refs [#2275](https://github.com/twisted/twisted/issues/2275)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_35"></a>|@therve commented|
|-|-|

shouldStop is used by stdlib TestResult, so it's used in trial. I added a test for it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun commented|
|-|-|

Looks good, please merge :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_39"></a>|@therve set status to `closed`|
|-|-|

(In [21138]) Merge trial-memory-2275

Author: therve
Reviewers: jml, exarkun
Fixes [#2275](https://github.com/twisted/twisted/issues/2275)

Add two optmizations to trial to reduce its memory footprint: add a new
DestructiveTestSuite, used by default, that don't keep a reference to test
instances after run, and replace the storage of successfull result by a counter.
This allows the garbage collector to clean a lot of objects, and also speed up
the whole suite.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
